File: patterns/adapter.js
//adapter pattern
//connect different interfaces

// Define the Target interface, which is the interface that the client expects to use
class Target {
    request() { }
}

// Define the Adaptee, which is the object that needs to be adapted
class Adaptee {
    specificRequest() {
        console.log("Adaptee: Specific request");
    }
}

// Define the Adapter, which adapts the Adaptee to the Target interface
class Adapter extends Target {
    constructor(adaptee) {
        super();
        this.adaptee = adaptee;
    }

    request() {
        console.log("Adapter: Adapted request");
        this.adaptee.specificRequest();
    }
}

// Use the Adapter to adapt the Adaptee to the Target interface
const adaptee = new Adaptee();
const adapter = new Adapter(adaptee);
adapter.request();

File: patterns/adapter.py
# Define the Target interface, which is the interface expected by the client code
class Target:
    def request(self):
        pass

# Define the Adaptee class, which is the class that needs to be adapted
class Adaptee:
    def specific_request(self):
        return "Adaptee: Doing specific request"

# Define the Adapter class, which adapts the Adaptee class to the Target interface
class Adapter(Target):
    def __init__(self, adaptee):
        self._adaptee = adaptee

    def request(self):
        return f"Adapter: {self._adaptee.specific_request()}"

# Use the Target interface to call the Adapter class, which in turn calls the Adaptee class
adaptee = Adaptee()
adapter = Adapter(adaptee)

print(adapter.request())
